modify Basic.__eq__ to detect dissimilar Numbers

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

closes #16916 

#### Brief description of what is fixed or changed

`Float(2) == 2` but the use of Float in an Expr will test as unequal `x**2.0 != x**2`.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- core
    - numbers still compare the same as they do in Python (`Float(1) == 1`) except when they appear in an Expression, e.g. `x**2.0 != x**2`
    - bugfix to `disambiguate` was made
- sets
    - `FiniteSet(1) == FiniteSet(1.0)`
<!-- END RELEASE NOTES -->

jksuom commented:
I think that taking types into account is a good idea. I wonder if that could be applied to all objects, not only numbers. So maybe something like this:

    elif type(a) != type(b):
        return False

oscarbenjamin commented:
This breaks the transitivity of `__eq__` for Tuple:
```julia
In [5]: Tuple(2) == Tuple(2.0)                                                                                                    
Out[5]: False

In [6]: Tuple(2) == (2,) == Tuple(2.0)                                                                                            
Out[6]: True

In [7]: Tuple(2) == (2,) == (2.0,) == Tuple(2.0)                                                                                  
Out[7]: True
```

oscarbenjamin commented:
And for FiniteSet
```julia
In [15]: FiniteSet(2) == {2} == {2.0} == FiniteSet(2.0)                                                                           
Out[15]: True

In [16]: FiniteSet(2) == FiniteSet(2.0)                                                                                           
Out[16]: False
```

jksuom commented:
Maybe it is `Expr.__eq__` that should be modified.

jksuom commented on sympy/core/expr.py:
```diff
@@ -101,6 +101,37 @@ def sort_key(self, order=None):
+        been defined by a class. See note about this in Basic.__eq__."""
+        return self._args
+
+    def __eq__(self, other)
```
:

smichr replied:
```suggestion
    def __eq__(self, other):
```

jksuom commented on sympy/core/expr.py:
```diff
@@ -101,6 +101,37 @@ def sort_key(self, order=None):
+        return self._args
+
+    def __eq__(self, other)
+        if type(a) != type(b):
```
`type(self) != type(other)`

smichr replied:
```suggestion
        if  not (self.is_Number and other.is_Number) and (type(self) != type(other)):
```

jksuom commented:
This looks promising. `FiniteSet(2) == FiniteSet(2.0)` still fails (as it does in current master) but that should change when `hash` has been fixed.

smichr commented:
> still fails (as it does in current master)

I just tried excluding numbers from the type test. Since FiniteSet is not an Expr, it should pass compare the items via number/number (Expr/Expr) comparison. I am thinking this will come to this routine, skip the type test and go to the hashable content test where a==b will pass.

smichr commented:
There are matrix/tensopr failures that need investigation:
```
___ sympy/matrices/expressions/tests/test_applyfunc.py:test_applyfunc_matrix ___
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/utilities/runtests.py", line 1308, in _timeout
    function()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/matrices/expressions/tests/test_applyfunc.py", line 59, in test_applyfunc_matrix
    assert expr.func(*expr.args) == expr
AssertionError
```

And this from tensor

```
_ sympy/tensor/array/tests/test_immutable_ndim_array.py:test_symbolic_indexing _
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/utilities/runtests.py", line 1308, in _timeout
    function()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sympy-1.5.dev0-py3.7.egg/sympy/tensor/array/tests/test_immutable_ndim_array.py", line 376, in test_symbolic_indexing
    assert Ms[i, j].diff(x) == ImmutableSparseNDimArray([[0, 3], [0, 0]])[i, j]

```

smichr committed:
modify Expr.__eq__ to detect dissimilar Numbers

smichr committed:
fix problem with disambiguate

smichr committed:
fix tests

smichr commented:
The first time the following is solved, the right answer is obtained in Py2. The next time, it is not:
```
>>> f,g=symbols('f g',cls=Function); eqn = Eq(Derivative(f(x), x), Derivative(g(x), x))
>>> dsolve(eqn,f(x))
Eq(f(x), C1 + g(x))
>>> f,g=symbols('f g',cls=Function); eqn = Eq(Derivative(f(x), x), Derivative(g(x), x))
Eq(f(x), C1 + Integral(diffx(g(x)), x))
```

smichr commented:
Does the `diffx` class in ode.py need to define `__hash__`? @asmeurer

oscarbenjamin commented:
> The next time, it is not

I can't reproduce that. Python 2.7 and SymPy master:
```julia
In [1]: f,g=symbols('f g',cls=Function); eqn = Eq(Derivative(f(x), x), Derivative(g(x), x))

In [2]: dsolve(eqn,f(x))
Out[2]: f(x) = C₁ + g(x)

In [3]: f,g=symbols('f g',cls=Function); eqn = Eq(Derivative(f(x), x), Derivative(g(x), x))

In [4]: dsolve(eqn,f(x))
Out[4]: f(x) = C₁ + g(x)

In [5]: dsolve(eqn,f(x))
Out[5]: f(x) = C₁ + g(x)
```

smichr commented:
Yes: not a problem in master, but a problem here. I don't know how to fix it and have just put the test in an `if PY3:` check. But that means the problem is still there and can show up on some other object...I just don't know how to tell who has this problem and why.

jksuom commented:
It seems that the local class `diffx`: https://github.com/sympy/sympy/blob/ab7dbc87d9d1dc8db0dfaf83fd131ee4edabf992/sympy/solvers/ode.py#L4106-L4129
is first introduced in `replace` and then removed in `unreplace` to make integration of a derivative possible. That is what happens in current master but in this PR `replace` and `unreplace` see different versions of `diffx`.

oscarbenjamin commented:
> in this PR `replace` and `unreplace` see different versions of `diffx`.

That makes sense. The diffx class is in the cached expression during a second call to dsolve which tried to unreplace a new diffx class. The simple solution is for diffx to be defined at module scope instead of in a function

smichr commented:
> simple solution is for diffx to be defined at module scope instead of in a function

I used to keep track of such details with common blocks and such with Fortran, but this level of thinking is not something I usually need for Python. Really appreciate the help on this.

smichr committed:
give diffx identification

In PY2, when the unreplace step is performed, the
diffx is not being found by replace. Although func_name
could be used to identify it, an attribute is being
used instead. This did not appear to be necessary in
pY3 so can (perhaps) be dropped when PY2 is dropped.

jksuom commented:
This seems to be working well.

smichr commented:
A way to make `Finite(2.0) == FiniteSet(2)` is to just use the sorted args as the hashable content instead of a frozenset of the same. Although `frozenset([1]) == frozenset([1.])`, `frozenset([S(1)]) != frozenset([S(1.)])`.

smichr committed:
FiniteSet(1) == FiniteSet(1.0)

This already works for Interval(1., 2) == Interval(1, 2).

smichr committed:
Update test_sets.py

smichr commented:
If all looks well, please feel free to commit. I have nothing further to add. 

This eliminates the irritating assymmetry of the former `x**2.==x**2` but `x**1. != x**1` -- now both (though for slightly different reasons) compare as unequal like `1.0*x != 1*x` and `2.0*x != 2*x`. Other than that, the impact has been much less than anticipated. I'm glad about that...and that my "mnemesis" `ode.py` was faced with a little help from friends. :-)

smichr commented:
Not sure where this error is coming from:
```consolve
Warning, treated as error:
duplicate object description of sympy.holonomic.holonomic, other instance in modules/holonomic/convert, use :noindex: for one of them
Makefile:49: recipe for target 'html' failed
make: *** [html] Error 2
```

smichr commented:
@sylee957 , do you have an idea of how to fix the Sphinx error?

sylee957 commented:
I think the build had started failing from the master after new update to sphinx
I've opened up an issue in #16945

oscarbenjamin commented on sympy/solvers/ode.py:
```diff
@@ -4102,12 +4103,13 @@ def _nth_algebraic_match(eq, func):
+            # don't use integrate here because fx has been replaced by _t
+            # in the equation; integrals will not be correct while solve
+            # is at work.
             return lambda expr: Integral(expr, var) + constant()
```
I guess it's not trivially possible to move diffx out of the function because it's using closed over variables...

smichr replied:
That was my experience, yes.

oscarbenjamin commented on sympy/solvers/ode.py:
```diff
@@ -4125,8 +4127,15 @@ def expand_diffx(*args):
-        return eq.replace(diffx, lambda e: Derivative(e, var))
+        if PY3:
+            return eq.replace(diffx, lambda e: Derivative(e, var))
+        else:
```
I don't understand why this is different between Python 2 and 3.

I added the diffx class and I now realise (seeing the work you've done) that it is fragile in the face of caching. That needs to be fixed and you've added a fix for Python 2 but I'm now worried about how safe it is for Python 3 as well. I'll look into it... (don't let that hold up this PR though)

asmeurer replied:
I've seen issues before with classes defined inside of functions. In this case, the class really does have to be defined inside the function, because it is different for each function call. Perhaps it would be better to store `var` and `constant` in the diffx.args. 

If I had to guess, the Python 2-ness of this points to the accursed class registry and the `__cmp__` (see https://github.com/sympy/sympy/pull/13059).

asmeurer replied:
Yep:

```
$ cat test.py
from sympy.core.core import all_classes

from sympy import Function, symbols, dsolve, Derivative, Eq

x = symbols('x')
f, g = symbols('f g', cls=Function)

eqn = Eq(Derivative(f(x), x), Derivative(g(x), x))

dsolve(eqn, f(x), hint='nth_algebraic')
dsolve(eqn, f(x), hint='nth_algebraic')

diffxs = [i for i in all_classes if 'diffx' in str(i)]
print(diffxs[0] == diffxs[1])
$ python3 test.py
False
$ python2 test.py
True
```

Because Python 2 uses the ill conceived [`__cmp__` method](https://github.com/sympy/sympy/blob/20234eb7782f1f66c1f1d2a6d05f1179091add0a/sympy/core/core.py#L73) of `BasicMeta` (`__cmp__` does nothing in Python 3).

jksuom commented:
This looks fine to me. The diffx issue is puzzling but I think that the implementation was questionable in view of caching. This PR only happened to bring that forth.

asmeurer commented:
Did you check if this introduces any performance regressions?
